### PR TITLE
NIFI-16045: Initial API for Processor / Connector Backlog

### PR DESCRIPTION
**Pull request by markap14**
# Summary

[NIFI-00000](https://issues.apache.org/jira/browse/NIFI-00000)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [ ] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [ ] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [ ] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [ ] Pull Request based on current revision of the `main` branch
- [ ] Pull Request refers to a feature branch with one commit containing changes

### CODE REVIEW

File: src/main/java/org/apache/nifi/components/Backlog.java [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+    }
+
+    /**
+     * Creates a Backlog whose only populated dimension is the {@code lastCaughtUp} timestamp.
+     * Useful for combining with a count-only Backlog via {@link #plus(Backlog)}.
+     *
+     * @param instant the moment at which the component was last observed as fully caught up
+     * @return a new Backlog
+     */
+    public static Backlog lastCaughtUp(final Instant instant) {
```
Comment (pvillard31) [inline src/main/java/org/apache/nifi/components/Backlog.java:264]: Since lastCaughtUp(Instant) is documented as useful for combining with a count-only Backlog, but plus() downgrades any count-only combine with it to AT_LEAST, is that downgrade intended when the timestamp adds no numeric uncertainty?

File: src/test/java/org/apache/nifi/components/TestBacklog.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    @Test
+    public void testPlusOverflowOnLongSum() {
+        final Backlog left = Backlog.flowFiles(Long.MAX_VALUE);
+        final Backlog right = Backlog.flowFiles(1L);
+        assertThrows(ArithmeticException.class, () -> left.plus(right));
+    }
+
+    @Test
+    public void testCaughtUpFactoryProducesZerosAndTimestamp() {
```
Comment (pvillard31) [inline src/test/java/org/apache/nifi/components/TestBacklog.java:124]: Can we add a small test for the builder rejecting negative flowFiles, bytes, and records values?

File: src/main/java/org/apache/nifi/components/connector/Connector.java [hunk 3]
```diff
@@ -242,4 +242,5 @@ public interface Connector {
      * @return a Future that will be completed when the draining is complete
      */
     CompletableFuture<Void> drainFlowFiles(FlowContext flowContext);
+
```
Comment (pvillard31) [inline src/main/java/org/apache/nifi/components/connector/Connector.java]: Is the extra blank line at the end of Connector intended, or a stray change?

**Review by pvillard31 — APPROVED** [review]
